engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/error/IsPayloadOfPanicTypeNode.java

```diff
@@ -0,0 +1,70 @@
+package org.enso.interpreter.node.expression.builtin.error;
+
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.interop.UnsupportedMessageException;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.Node;
+import com.oracle.truffle.api.nodes.NodeInfo;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import org.enso.interpreter.node.expression.builtin.meta.IsSameObjectNode;
+import org.enso.interpreter.node.expression.builtin.meta.TypeOfNode;
+import org.enso.interpreter.runtime.EnsoContext;
+import org.enso.interpreter.runtime.builtin.Builtins;
+import org.enso.interpreter.runtime.callable.atom.AtomConstructor;
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.type.TypesGen;
+
+/** An implementation of the payload check against the expected panic type. */
+@NodeInfo(shortName = "IsPayloadOfPanicTypeNode")
+public abstract class IsPayloadOfPanicTypeNode extends Node {
+  private @Child IsSameObjectNode isSameObject = IsSameObjectNode.build();
```

Comment:
Should this logic be the same as in [Meta.is_a](https://github.com/enso-org/enso/pull/3949) check and `case_of_on_type` check? Shouldn't we have a single `IsATypeNode` and share it among all the implementations?

Reply:
It's a subset I think. Since (I think?) I don't care about things such as strings, arrays, numbers, dates etc

Reply:
I agree with @JaroslavTulach that it seems best to have a single implementation for these.

@hubertp why do you think it is a subset? Does the `IsPayloadOfPanicTypeNode` not check for strings and numbers?

Will `Panic.catch Integer (Panic.throw 42)` work?  I'm not certain that it has to, but it used to work.

Surely, as long as `Panic.throw` will accept a type of payload, the same type of payload must be catchable by `Panic.catch`.

So if we want to disallow catching integers, we need to disallow throwing them as well IMO.

Reply:
Yes, I knew that there will be some intersection with `Meta.is_a` builtin work by @JaroslavTulach but I didn't want this to be a blocker.
And yes, the limitation was on purpose. But it seems that `Panic.throw` doesn't make that distinction so I should relax indeed (feels wrong). 

So should I wait on `Meta.is_a` work to get in first? @JaroslavTulach any ETA when will this get in?

Reply:
Feels wrong to catch an Integer? I guess we could disallow throwing it too. But not sure where to draw the line? Just disallow primitive types (Text, Number, maybe Array)?

Reply:
> Feels wrong to catch an Integer? I guess we could disallow throwing it too. But not sure where to draw the line? Just disallow primitive types (Text, Number, maybe Array)?

Similarly to Java/Scala I would expect the value to be thrown to be of a known base type (there `Throwable`). It does not appear like we use this in our tests but surprisingly we have an example with that in `Panic.throw`. We should certainly have tests for things that we have in the _example_ section.

Reply:
I've been looking at this piece of code and actually I was wrong about 

> It's a subset, I think. Since I don't care about things such as strings, arrays, numbers, dates etc

Since the code is very similar to `CatchTypeBranchNode` it is only polyglot arrays.
Also looking at the code again, I discovered another bug in the existing code of
```
case Integer.valueOf "1" of
    _ : Any -> 0
    _ -> 1
```
would return `1` rather than `0`. I fixed that in `catch` but not in the `case` branch check.

So I will try to handle polyglot arrays (since I guess one could throw them, for whatever reason) and port a fix for the above. But I would delay merging with `Meta.is_a` when it actually lands.

Reply:
So `CatchTypeBranchNode` is now merged with `IsValueOfTypeNode` so one less thing to deal with.

Reply:
> the value to be thrown to be of a known base type (there Throwable)

There is also the _"JavaScript way"_ - you can throw anything - strings, numbers, etc.

Reply:
> So `CatchTypeBranchNode` is now merged with `IsValueOfTypeNode` so one less thing to deal with.

OK, so since 61d23a44e0a1e4ead6ad79544506be28d35760b9 the _type checking_ code shall live in ``IsValueOfTypeNode` and be shared. I'll count with that in my #3949.

Reply:
> > the value to be thrown to be of a known base type (there Throwable)
> 
> There is also the _"JavaScript way"_ - you can throw anything - strings, numbers, etc.

Very good point!

```
from Standard.Base import all

foreign js foo = """
    return 42+23;

foreign js bar = """
    throw 42;

main =
    IO.println foo
    r = Panic.catch Any bar err->
        IO.println err
        "error handled"
    IO.println r
```

```
65
(Panic (Polyglot_Error.Error 42) 42)
error handled
```

Our current `Panic.catch` is capable of catching JS exceptions, so probably would be good to keep that and ideally would be great if I could also do `Panic.catch Integer` there - currently it does not work due to `Polyglot_Error` wrapper - I did not check if on your branch it now works.

Either way - if possible it would be great to add a test case to `Error_Spec` checking catching the JS exception - and if possible also checking there if filtering like `Panic.catch Integer js_function_throwing_js_int` works.

Reply:
> Either way - if possible it would be great to add a test case to `Error_Spec` checking catching the JS exception - and if possible also checking there if filtering like `Panic.catch Integer js_function_throwing_js_int` works.

This proved to be a bit more painful than I initially imagined but that is now possible (as checked by additional tests).